File: find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
                /*
            - we need to check if array is already sorted
            if (nums[0] < nums[N-1]) return nums[0]
            
            - we need to check if the array length is 1
            if (N == 1) return nums[0]
            
            - Modified binary search:
            
            1) checking nums[mid] < nums[mid + 1] 
                return nums[mid + 1]
            
            2) checking nums[mid - 1] < nums[mid]
                return nums[mid]
                
            Updating our pointers. 
            
            if (nums[mid] < nums[0]) 
                high = mid - 1
            else
                low = mid + 1            
        */
        
        int low = 0;
        int high = nums.length - 1;
        
        if (nums[0] < nums[nums.length - 1]) {
            return nums[0];
        }
        
        if (nums.length == 1) {
            return nums[0];
        }
        
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] > nums[mid + 1]) {
                return nums[mid + 1];
            }
            if (nums[mid - 1] > nums[mid]) {
                return nums[mid];
            }
            if (nums[mid] > nums[0]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }
}

File: find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # check if array is already sorted
        if len(nums) == 1:
            return nums[0]
        
        if nums[0] < nums[len(nums) - 1]:
            return nums[0]
        
        low = 0
        high = len(nums) - 1
        
        while low < high:
            pivot = (high + low) // 2
            
            # [1,2,3,7,8]
            
            # LHS is increasing && RHS is decreasing
            if nums[pivot] > nums[pivot + 1]:
                return nums[pivot + 1]
                
            # RHS is increasing && LHS is decreasing
            if nums[pivot] < nums[pivot - 1]:
                return nums[pivot]
        
            # move pointers
            if nums[low] > nums[pivot]:
                high = pivot - 1
            else:
                low = pivot + 1
        
        return -1